imagebuildah/stage_executor.go

```diff
@@ -1193,68 +1134,34 @@ func (s *StageExecutor) tagExistingImage(ctx context.Context, cacheID, output st
 		if err != nil {
 			return "", errors.Wrapf(err, "error getting history of base image %q", s.builder.FromImageID)
 		}
 	}
-	for _, image := range images {
-		var imageTopLayer *storage.Layer
-		var imageParentLayerID string
-		if image.TopLayer != "" {
-			imageTopLayer, err = s.executor.store.Layer(image.TopLayer)
-			if err != nil {
-				return "", errors.Wrapf(err, "error getting top layer info")
-			}
-			// Figure out which layer from this image we should
-			// compare our container's base layer to.
-			imageParentLayerID = imageTopLayer.ID
-			// If we haven't added a layer here, then our base
-			// layer should be the same as the image's layer.  If
-			// did add a layer, then our base layer should be the
-			// same as the parent of the image's layer.
-			if buildAddsLayer {
-				imageParentLayerID = imageTopLayer.Parent
-			}
-		}
-		// If the parent of the top layer of an image is equal to the current build image's top layer,
-		// it means that this image is potentially a cached intermediate image from a previous
-		// build.
-		if s.builder.TopLayer != imageParentLayerID {
-			continue
-		}
-		// Next we double check that the history of this image is equivalent to the previous
-		// lines in the Dockerfile up till the point we are at in the build.
-		manifestType, history, diffIDs, err := s.executor.getImageTypeAndHistoryAndDiffIDs(ctx, image.ID)
-		if err != nil {
-			// It's possible that this image is for another architecture, which results
-			// in a custom-crafted error message that we'd have to use substring matching
-			// to recognize.  Instead, ignore the image.
-			logrus.Debugf("error getting history of %q (%v), ignoring it", image.ID, err)
-			continue
-		}
-		// If this candidate isn't of the type that we're building, then it may have lost
-		// some format-specific information that a building-without-cache run wouldn't lose.
-		if manifestType != s.executor.outputFormat {
-			continue
-		}
-		// children + currNode is the point of the Dockerfile we are currently at.
-		if s.historyAndDiffIDsMatch(baseHistory, baseDiffIDs, currNode, history, diffIDs, addedContentDigest, buildAddsLayer) {
-			return image.ID, nil
-		}
+
+	layerKey := cache.CalculateBuildLayerKey(s.executor.outputFormat, buildAddsLayer, parentLayerID, nextCreatedBy, baseHistory, baseDiffIDs)
+	return layerKey, nil
+}
+
+// cachedIntermediateImageExists returns the layer image id if exists
+func (s *StageExecutor) cachedIntermediateImageExists(ctx context.Context, buildAddsLayer bool, parentLayerID string, nextCreatedBy string) (string, error) {
+	layerKey, err := s.currentLayerKey(ctx, buildAddsLayer, parentLayerID, nextCreatedBy)
+	if err != nil {
+		return "", errors.Wrap(err, "error obtaining the layer key")
+	}
+
+	cacheID, err := s.executor.localLayerProvider.Load(layerKey)
+	if err != nil {
+		return "", errors.Wrap(err, "error obtaining the layer")
```

Comment:
Please do not start error messages with "error"

Reply:
resolved